### DISCUSSION TITLE
AFE using MCP6002 op-amp - Design #5

### DISCUSSION THREAD

**Discussion by fhdm-dev** [fhdm-dev/scoppy#63] [Scoppy Front end designs]
Here's a super simple analog front end that uses the (cheap and available) MCP6002 R-to-R dual op-amp. This particular design uses one of the amps (and channels) for a +-7V input range and the other for a +-1V range. The design also incorporates under/overvoltage protection.

The schematic can be viewed in EasyEDA (no login required) [here](https://easyeda.com/editor#id=5d5f637d6a8448b59f484372ed679191).

The 3 resistors on the input ensure the appropriate attenuation and also adds an offset. 1N4148 diodes on the non-inverting input ensure the voltage is clamped between approximately -0.55V and +3.85V which is well within the MCP6002's allowable range of Vss-1V and Vdd+1V.

Channel 2 has two extra resistors. These control the gain of the op-amp. Channel one is in unity gain configuration and so doesn't need these resistors.

There is a 100 ohm resistor on the input of each ADC pins. These serve two purposes:
1. To stabilize the op-amp (the ADC is a capacitive load - though small)
2. To limit the current into the ADC in case of an unexpected over-voltage event

If you want to use this design but using different input voltage ranges then see [here](https://easyeda.com/editor#id=7d1d63aa4eb740938d0ed4895d41cec6) for the appropriate resistor values to use.

If building this design on a solderless breadboard the signal will probably appear quite noisy in the app. It should work better on prototyping board especially if the wires connected to the non-inverting input of the op-amp are kept as short as possible (this part of the circuit is very susceptable to picking up interference).
       
Once you've built the AFE you'll need to update the _Voltage range_ settings in the app. To do this, tap the channel badge at the bottom of the screen, then tap _Settings_ and then _Voltage ranges_. The values to enter can be found on the [schematic](https://easyeda.com/editor#id=5d5f637d6a8448b59f484372ed679191). If you're fussy, then you can even do a [calibration](https://oscilloscope.fhdm.xyz/wiki/Vertical-Calibration) so that the voltages shown in the app are more accurate.
       
![afe](https://user-images.githubusercontent.com/52391579/174912454-19ab595d-eabd-477f-82d8-50a163dac2dc.jpg)
Two channel AFE on proto board. Just add MCP6002 op-amp.
   
![scope](https://user-images.githubusercontent.com/52391579/174912584-056eced0-f1bc-4d36-8f70-f12cc540e6ca.jpg)
3.3Vpp square wave on CH1
0.7Vpp sine wave on CH2
   
![fft](https://user-images.githubusercontent.com/52391579/174912649-82bc8e03-1e1d-4780-822a-b07031e43076.jpg)
FFT on this cheapskate oscilloscope

![pico-attached](https://raw.githubusercontent.com/fhdm-dev/scoppy/main/images/diyscope-phone-sine-square-1024w.jpg)
.... and with the Pico attached to the board with headers. Looks much nicer.

**Comment by paprika27** [comment]
Hi FHDM-dev, I just ordered the MCP6292 and RPu - I saw you pointed out for Design 1 how to change the resistors for 10x probes. Could you kindly point out for me, what I would need to change with this design? Many thanks for your awesome work!

**Reply by fhdm-dev** [reply]
Hi,

Interesting question because I didn't consider 10x probes when doing this design.
However, if you change RX_1 to 470k then the input range on channel 1 will be -66.5V to 69.8V with 10x probes - not quite 10x attenuation but close. Or do you still want the input range to be +-7V with the 10X probes?

**Reply by paprika27** [reply]
thanks so much for your speedy reply! I basically looked at "this is the design we recommend" and am now thinking how to use my existing probes from my non storing analog scope through a BNC connector for those cases where I need storage (transients) and FFT mostly for ripple and hum. I currently attach a speaker, capture with my phone mic and use another android spectrometer to achieve this. So I'm not sure what input range I want - with my limited understanding, changing it to -66.5 to 69.8 in the frontend but leaving the app to interpret -6.65 to 6.98 should basically be x10? Accuracy is not critical.

**Reply by paprika27** [reply]
I feel like my understanding is indeed very limited... if I measure a 20v signal with my x10 probe, it should give me 2v which would be within spec for the default frontend desing. On the contrary, I would want to change the app to interpret these 2v as the 20v they actually are, so I should wrongfully configure its input range to 10x what it is..

**Reply by fhdm-dev** [reply]
My fault. I didn't explain it very well.

This particular AFE design doesn't work well if you want to switch your probes between 10X and 1X. For example, if you've chosen resistor values such that the range is +-7V when using 1X probes then it doesn't automatically become +-70V when you switch to 10X - like you would expect with a real scope frontend (it actually becomes -142V to +70V).

So when calculating the resistor values for a channel you need to know in advance whether you'll be using 10X probes or 1X. If you want to switch between 10X and 1X (and have the correct values displaying in the app) then you'll need to use a different AFE design.

Regarding your example above about 20v with 10X probes being interpreted as 2v by the app, there is a setting in the app where you can set the probe attenuation so that the voltage will be displayed correctly (tap the channel badge at the bottom of the screen and then tap settings). But this is really only useful for front ends that do support switching between 10X and 1X. 

I don't think I explained that very well again :frowning_face: 

Anyway, if you let me know the minimum and maximum voltage ranges you'll be measuring I can let you know what resistor values to use.

**Reply by paprika27** [reply]
Thanks so much for explaining your design and how this stuff works! My old probes are 10x without switching - so I think I would want to change the frontend for that (permanently). If I understand you correctly, the default frontend range of +/- 7V with 10x probes and setting the range +/- 70V would enable me to measure 70V signal and get it displayed as roughly 70V.
The largest I have had on my bench was 208 VDC (tube amp) but I've been doing that with 100x probes on my analog scope and said workaround for FFT. Maybe at some point I understand enough to configure one channel for each attenuation. I think I'll need storage for digital circuits so maybe in the range +/- 32V (that's what my bench PSU can do ;) ). So if I have understood things correctly, using 10x probes on the default input range +/- 7V with wrongfully configured app (set to +/- 70V?), I should get close enough without breaking things.

**Reply by fhdm-dev** [reply]
That makes sense, though I think you should ignore the fact that there is a +/- 7V range on the schematic. You can choose whatever range you like and calculate the appropriate resistor values for that. Once you've entered the actual voltage range into the app it will show the correct value (ie. you don't need to wrongly configure it and you don't need to change the probe attenuation settings in the app). And for all the ranges you can choose to use either 10X or 1X probes when calculating the resistor values - you just can't switch between them.

Here's some examples:
approx. +/- 70V range using 10X probes:
R1_1=470k
R2_1=470k
RX_1=470k
Actual Input Range Min=-66.5V
Actual Input Range Max=+69.8V

+/- 70V range using 1X probes:
R1_1=1000k
R2_1=47k
RX_1=47k
Actual Input Range Min=-70.2V
Actual Input Range Max=+73.5V

+/- 32V range using 10X probes:
R1_1=220k
R2_1=1000k
RX_1=910k
Actual Input Range Min=-33.4V
Actual Input Range Max=+33.7V

+/- 32V range using 1X probes:
R1_1=1000k
R2_1=100k
RX_1=100k
Actual Input Range Min=-33.0V
Actual Input Range Max=+36.3V

In the examples above the 'Actual Input Range' values are the ones you enter into the app as 'Voltage range(s)'.

As for not breaking things, I've never worked with voltages that high so I can't guarantee anything.
All I do know is that applying 18+V to the input of the MCP6292 without limiting the current will destroy it. I learnt that the hard way :smile: 
So I would make sure that the attenuation and overvoltage protection is working correctly before attaching the MCP6292. That is, the voltage at the inputs of the op-amp shouldn't ever go below -1V or above 4.3V.

**Reply by paprika27** [reply]
Thank you so much! Your dedication and support are truly remarkable, thank you! I'm now eagerly awaiting the arrival of the MCPs.

**Reply by paprika27** [reply]
I finally found a free evening and it's working! I'm very excited about this. I tore down ch2 when leaving it unterminated seemed to obfuscate ch1 - I can't get play store to work on this old spare so I couldn't use it, anyway. Your input ranges seem to allow me to match what my uncalibrated old analog scope shows and the probe selection works as I would have hoped (I didn't add a bnc port for my 10x probes, yet). There seems to be a little bit of clipping in the troughs of the signal but nothing I worry about yet. I'm very glad FFT includes a simple V scale - with my noisy breadboard and jumper wires, the relative scale obfuscates what I'm looking for. Thanks again for the great instructions, the great apps, the great support! Let me know where you would like to see this make posted.


![IMG_20220729_234200](https://user-images.githubusercontent.com/8465938/181852263-1b4a3907-311d-4aa8-964c-b93f42218c60.jpg)
![IMG_20220729_234213](https://user-images.githubusercontent.com/8465938/181852269-10c05a91-bdd7-4ac3-95be-7bbfca788e61.jpg)
![IMG_20220729_235127](https://user-images.githubusercontent.com/8465938/181852256-ad00ea4b-4ab9-4b8c-b9f4-bbc9253a2cfe.jpg)![IMG_20220729_234236](https://user-images.githubusercontent.com/8465938/181852253-eb2c5181-9e1c-4767-9d87-657b0dbb6c20.jpg)
![IMG_20220729_234227](https://user-images.githubusercontent.com/8465938/181852245-07c21ceb-9af4-4b58-b5ac-39ee30228207.jpg)
!

**Reply by fhdm-dev** [reply]
Nice :)
I think instructables or hackster.io are good places to post projects like this. I would steer well clear of hackday.
Pity the play store didn't work. That advertising banner is quite ugly! I'm not sure what the problem is but I can send you a promo code for the premium upgrade if you like. Maybe that would work?

**Reply by dagelf** [reply]
Hack-a-day is great, it's just that they're just very anti-commercial, I presume because most people who go there work for big companies and they go to escape and do something hobbyist... so they don't want it to become just another site filled with ads for commercial products like their day job... or many electronics magazines. The fact that your project was mentioned there is a small miracle, and great! But I think its because of your interactions on github which is very much in the FOSS spirit. Their readership might not understand the difference between big and small projects... :-) Anyways, thanks again for building and sharing something so useful!

**Comment by Dupie123** [comment]
good day. i have had a look at the 5 designs you offer for the scope and i was wondering if it would be possible to build a scope with 2 inputs like in design number 5 but combine that with design number 4 where i can select between multiple voltage ranges for each channel. i am quite new to circuit design and id rather ask advise before making mistakes

**Reply by fhdm-dev** [reply]
Hi,

I think you should base your design on Design 4 (and just duplicate the circuitry for the extra channel). The arrangement of the resistors on the input of design 5 means that you'd need to switch at least 2 resistor values when changing voltage ranges.
   
If you want to use an MCP6002 (or equivalent) op-amp instead of the LM358 with Design 4 then you'll need to make the following changes:
- power the op-amp with 3.3V rather than 5V (VSYS)
- add a extra diode to the +ve input of the op-amp for overvoltage protection (1N4148 diodes can be used here)
- remove the 2nd op-amp (U1B) and diode D2. Over/under voltage protection isn't needed here because the output of the MCP6002 won't swing outside of the 0-3.3V range.

This will only work for positive input voltages. If you want the input voltage range to include negative voltages you'll need to provide an offset voltage (see Design 3. This is labelled Vref). The offset voltage will also need to change whenever the voltage range is switched.

An alternative to using a mechanical switch to change the ranges is to use an analog switch like a 74HC4052. The firmware can then be configured so that the appropriate range is selected when you change the Volts/Div in the app. However, I haven't documented how to do this except for some comments in the scoppy-local-example.cmake file in the firmware. A hardware design using this method can be seen [here at OSHWLab](https://oshwlab.com/fruitloop57/fscope-250k5-v2_copy)

**Reply by ema987654** [reply]
Hi, 

  I'm really interested in this design idea. I'd love to base my design from AFE4, using the MCP6002 on each channel, but also being able to work with negative input voltages.
  The issue I find is I don't really get how should I change the offset voltage on each range. I figured I can use one of the free Op Amps on each 6002 to get the VREF value, but not as far as knowing what should I change on each range, for this I plan on using a 3 poles 4 throws selector for this. 

  Thanks for contributing and I'm looking forward to your answer!

**Reply by fhdm-dev** [reply]
Hi,

If you look at AFE3 (which does do negative voltages), you could use one pole to switch the value of Rref1 to give you the correct value of Vref. Another pole could be used to switch the value of Rf2 so you get the required gain. And the 3rd pole can be connected to the voltage   range pins on the Pico so that the firmware/app knows which range you've selected.

The fun part is calculating the resistor values you'd need for each of your input voltage ranges, then recalculating using resistor values that are actually available and then working backwards to find out the voltage ranges you'll actually end up with. I you need help with this then let me know.

I just noticed AFE3 has very high values for the Rf2 and Rg2 resistors. I would use something more like 10K (or lower) for Rg2.

I hope that helps :smile:

**Reply by ema987654** [reply]
Hi, 
Yup, this was really helpful, I was scratching my head figuring out how AF3 works yesterday and kind of figured it out. But just like you said the fun part is to use the right voltage divider / gain, I think I'll go ahead and make a design based off all this info, and post back with a diagram as soon as I get it done in terms of calculations while I test it out, so if someone out there had the same idea can test/ improve upon it.

Thanks and TGIF.

**Reply by ema987654** [reply]
Hi,
![image](https://user-images.githubusercontent.com/55396753/186797101-8db10024-f304-41ca-bf1b-e6bc78cdba9f.png)
I've been at this design for a week. I've made some LTSpice simulations to make sure my calculations were right at least on some values.
  Anyway, I've chosen the following ranges:

*Switch on position 1 (VRange LOW LOW):
Vmin=-3,43V; Vmax=3,51V
 *Switch on position 2 (VRange LOW HIGH):
Vmin=-5,21V; Vmax=5,15V
*Switch on position 3 (VRange HIGH LOW):
Vmin=-12,6V; Vmax=12,46V
*Switch on position 4 (VRange HIGH HIGH):
Vmin=-32,96; Vmax=33,04V

I went with the MCP6002 for the faster slew rate over the LM358 (at leas typical values). Feel free to criticise my design I'd love some input specially on the +-33V range input since is 10x the ADC voltage and honestly freaks me out a little bit, as well as the Schottky diodes  for protection (I'm certainly not an expert).
Also I wasn't sure where I should use the selector for the least amount of noise (I think open wires from the PCB to the selector can act as a nice antenna) so that's why I picked all the resistors in parallel straight out from VRef or GND it might be the worst choice (as I said, certainly not an expert).

Thanks for your help and I hope my design based of the AFE3, AFE4 and AFE5 comes in handy for everyone here.

Cheers.

**Reply by fhdm-dev** [reply]
Looks good. I think the 33V should be fine as long as your 3.3V rail can sink enough current if you accidently have the selector set to the wrong range (or worse - if it's in between ranges while the switch is being rotated).

Regarding noise. You probably only need to worry about the high impedance part of the circuit (ie the non-inv input of the op-amp) - so I'd try and make the wires on that pole as short as possible. I'm not sure what else you can do about noise apart from adding some shielding.

And thanks for sharing. I'm sure others will find this very helpful.

**Reply by ema987654** [reply]
Hi,
Thanks for your reply. So basically the design is fine, the only thing I'd probably change (and realized that once I was too deep into this design) is to avoid feeding the signal directly to the non inverting input of U1B. Is one of those things that I can't help to fix, otherwise if is intended to work with up to +-33V and the selector is rotated, there would be no voltage divider on the input and we fully depend on those sink Schottky diodes along with the 680K resistor in series to protect the MCP6002.
Thanks for your input!

**Reply by SteveTinkers** [reply]
Nice Idea!  Anything wrong with just having 3 or 4 inputs per channel and just plugging in?  

I like the switch idea too. @ema987654 how did it turn out?  What switch did you end up using?

**Reply by ema987654** [reply]
@SteveTinkers Sorry for the late reply. It worked great, I'm actually trying some signal compensation circuits to counter the ADC input capacity, but the circuit itself worked great. I have an experimental PCB, the sort that ends up being a working prototype and perhaps you never built the real thing given I have already designed and 3d printed a case LOL. The switch I've used is a rotary type, 4 positions, 3 poles.
![IMG_20221212_162128_HDR](https://user-images.githubusercontent.com/55396753/207135746-8891e656-af5c-4d48-88ef-cf3db9bd4b40.jpg)
![IMG_20221201_183846](https://user-images.githubusercontent.com/55396753/207135791-b7f5b479-7d61-4b78-93c4-dd7da7240c02.jpg)
BTW I'm from Argentina, giving you the link of where I got the switch would be pointless.

**Reply by MeisyarohAzzahra** [reply]
May I know the size of the black casing for the circuit holder? @ema987654 

https://user-images.githubusercontent.com/55396753/207135791-b7f5b479-7d61-4b78-93c4-dd7da7240c02.jpg

**Comment by krisstakos** [comment]
![Screenshot_20220827-230412](https://user-images.githubusercontent.com/5090779/187046399-7a243f7b-fad1-41bd-8361-890f567ea046.jpg)
Hey, @fhdm-dev  nice work. I put it together on a prototype board as tight as possible with the only differences that the channel 2 is in unity gain configuration and the resistors R1_1, R2_2 and RX_2  are configured for (-10.2V; +10.1V). Andd I'm not sure what I'm missing. I can see the waveforms on both channels but the voltages are all wrong. Tried to do calibration according to your docs but without luck. Also I've checked my connection many times and they are all alright

On the screenshot I'm using GP22 to generate 1khz signal

**Reply by krisstakos** [reply]
I just assembled it on the breadboard and the result is absolutely the same. For 3,3v square-wave I always see Vmax around 1,4-5V Is it not properly calibrated or my op-amps are bad?

**Reply by krisstakos** [reply]
Tried the same circuit but with LM358, same voltage output, seems like it's not from the op-amp. Any advice?

**Reply by fhdm-dev** [reply]
Hi,

You could try this.
Connect the CH2 input to GND. The voltage at the non-inverting input of the op-amp should be approximately 1.65V. The voltage at the output of the op-amp should also be approximately 1.65V.
Then apply 9V DC at the CH2 input. The voltages should be approximately 3.14V.

If the voltages are wrong then that suggests a problem with the circuit (maybe wrong wiring or wrong resistor values).
If the voltages are OK then the problem is likely to be the input voltage range setting in the app.

**Reply by krisstakos** [reply]
@fhdm-dev  these are my results, based on this maybe I should check my connections

> CH1 to GND
VINA+: 1.704
OUTPUT: 1.721

>CH2 to GND
VINB+: 1.651
OUTPUT: 1.666

>CH2 INPUT 9.88V
VINB+: 3.221
OUTPUT: 2.054

**Reply by fhdm-dev** [reply]
OK. The first two look fine.
With the last measurement the VINB+ value is what I would expect but the output obviously isn't right given that you've configured the amp with a gain of 1. That suggests to me that the op-amp you are using isn't able to output a voltage as high as 3.3V. This would happen for example with an LM358 if the supply voltage is 3.3V (because it can't swing the output close to the positive supply rail). The MCP6002 should be OK though because it's output is able to swing within 25mV of the supply rails - so I'm not sure why that wasn't working :thinking:

**Reply by krisstakos** [reply]
Maybe because I ordered them from aliexpress...

**Reply by krisstakos** [reply]
In theory, if I apply more supply voltage let's say 5v it should get closer to 3.3v ?

**Reply by fhdm-dev** [reply]
Yes, just make sure you have a current limiting resistor between the op-amp and the ADC inputs because the voltage could end up higher than 3.3V.

**Reply by krisstakos** [reply]
Thanks @fhdm-dev, seems like I got scammed with these op-amps. Applying 5v Vdd seems to do the job at least for simple voltage follower configuration.

**Reply by n9jcv** [reply]
I ordered mcp6002s from Ali and they were not rail to rail.  Definitely fakes.

**Reply by zoltanprociow** [reply]
hey. i have same problem with mcp6002 but there is microchip logo on it, are sure its a problem of manufacture? do you know what would be good replacment ? or what are the  requirements for opamp.

**Reply by fhdm-dev** [reply]
The basic requirements are that it should work with a power supply of 3.3v, it should be rail-to-rail, have a GBW of at least 500kHz and be unity gain stable.

**Reply by zoltanprociow** [reply]
I got mcp6004 so I think to make 4 channels and swiching between them. And  I heve noob questions:
- does the input resistance has to be so high, like in yours examples , or it doesn't matter. In the lowest of my configurations there is: R1 2,2k  R2 1,2k Rx 750R = -9,68 +9,35  input resistence 2,66
- does the LOW state of GIPO2/3 and GIPO4/5 has to be connected to GND ?
- can I send signal from opamop to gio26/27 through cd4053 used as swich between channels, or it can have substancional influence on measured signal?

**Reply by fhdm-dev** [reply]
Nice. The MCP6004 should work fine :smile: 

Most oscilloscopes have an input resistance of 1M to prevent loading of the circuit being measured (that is, to prevent current flowing into the oscilloscope front-end and affecting the circuit under test). Having a high input resistance allows the oscilloscope to be able to accurately measure voltages when the output resistance of the circuit under test is relatively high. You don't want a large voltage drop across the output resistance of the circuit you are measuring because that would make the voltages shown in the oscilloscope too low.

You could use lower values than those I've shown in the examples but a total input resistance of 2,66k is (IMHO) far too low. Hundreds of K is what I would aim for as a minimum. However, a higher input resistance will make the signal appear more noisy so you just need to get the balance right for your needs.

GPIO2/3/4/5 all have pull-down resistors enabled (internal to the RP2040 IC). If they are not connected they will always read low.

I believe a cd4053 would work. I normally use the 74HC versions of these switches because of the lower on-resistance. It will introduce some distortion but probably not enought to worry about.

**Comment by krisstakos** [comment]
Hi again @fhdm-dev how you've calculated the RX resistor values also what's the purpose of it? It will be very helpful if we got any explanation in the docs, formulas even better.

**Reply by fhdm-dev** [reply]
The purpose is to push the voltage higher at the non-inverting input of the op-amp so that the bottom of our input range (eg. -10V) becomes 0V at IN+. We need to do this because the Pico ADC only works with in the range of 0V to 3.3V.

When the op-amp has unity gain, the value of Rx is the parallel combination of R1 and R2.

The same thing can be achieved by using a voltage divider to create a reference voltage at the base of R2. This method is used by AFE3 (shown as Vref). This is actually easier to understand but requires an extra resistor. In AFE3, the op-amp (U1D) is probably not necessary if the values of Rref1 and Rref2 are low enough. The calculations for this method can be found in a TI document called 'Designing Gain and Offset in Thirty Seconds'.

I'll see what I can do about an explanation and forumulas etc.

**Reply by krisstakos** [reply]
Thanks!

**Reply by SteveTinkers** [reply]
Double Thanks!  This is a cool project!  

To confirm, RX and RG resisters are only needed for Channel 2? And apparently only for voltages below 3.3, because you only list values for max 3v, 2v, and 1v configurations.  Interestingly, resistance jumps up on RF for 1v range...it was trending down.
[![AFE5-Best-Resisters.png](https://i.postimg.cc/YSLwXRdS/AFE5-Best-Resisters.png)](https://postimg.cc/Mcw4H72k)

Can you help me calculate resister values for a  14-15 volt system?  I'm trying to analyze some older automotive busses that float 12v.  14 or 15 is close enough to be useful for lots of things... and sort of in between 6.6 and 36!
What does RX do?  I calculated R1 3600, R2 1000 should be good to 15.1V.  Agree?  It doesn't fit the pattern....

Bonus round dumb question... any reason why you went from one cap in your pictures to two (100nf and 10uf) in your diagram?

**Reply by fhdm-dev** [reply]
I assume you mean RF and RG. These are only needed for voltage ranges where the gain of the op-amp isn't unity. So yes, in the linked schematic they are only needed for channel 2 because for channel 1 the op-amp does have a gain of one.

Even though the value of RF doesn't trend up or down consistently, the ratio of RF to RG does  (ie. the gain does).

RX is used to raise the voltage level so that a negative voltage applied to the 'scope becomes a positive voltage at the Pico ADC . If your input range doesn't include negative voltages then this isn't needed.

Yes, I agree that R1=3600 and R2=1000 will give a range of 0V to 15.1V. No RX is needed.

Regarding the capacitors, the MCP6002 datasheet suggests using two bypass capacitors which is why you see two in the schematic. However when I built the circuit I was lazy and didn't add the 2nd one - oops :smile:

**Reply by SteveTinkers** [reply]
Makes sense! 

So It's a simple standard voltage divider if you dont need negative voltages.  How can I incorporate an RX in 15.1 circuit?  Just to make things consistent so I can use about -15.1 too.

I was looking at the above switch idea, it's interesting but add complexity and one more part.  Anything wrong with just having 3 or 4 inputs per channel and just plugging in the needed range?  I've thought about soldering everything together (because you mention permanent  installation on a breadboard = noise) and keep the resistors swappable.  I'd probably use 6V or 15V most often.  If I find something is not detailed enough for lower voltage (1V/3V), swap em. What do you think?

1V, 6V, 15V, 30V (or even straight to 70) would cover LOTS of options.  I'm assuming as the range gets higher it gets less accurate.  I debate having a 3.3v direct to the diodes...

**Reply by fhdm-dev** [reply]
R1=680, R2=180,RX=150 will give you a range of -15.0 to +15.8

Yep, you could have multiple inputs per channel. If you want the correct voltage shown in the app you need to apply the appropriate logic level to the [voltage range GPIOs](https://oscilloscope.fhdm.xyz/wiki/Analog-Front-End) so that the app knows which range you are using.

You are right, as the range gets larger the accuracy decreases. As this is an 8-bit scope the vertical resolution is the range divided by 256.
eg. for the -15 to 15.8 range it would be 93mV.

**Reply by SteveTinkers** [reply]
Thanks!  What calculation is that?  Can I learn how you determined those resisters? I dont think I'll ever need to change them.  Not sure why I'd use 11 v or 15 but I like to learn!

I saw the voltage range GPIOs.  I didn't understand what range 0, 1, 2, 3 meant.  Are those actual voltage.  Choose range 3 for anything above 3 volts?  I was hoping it'd make sense once I built it and had the app. 

Do you split up all the pins to the RasPi for  a reason?  They are all so far from each other.  Can I keep them together and make a header cable out of some 4 pin jack?  Does putting them together make more noise or something?
> You probably only need to worry about the high impedance part of the circuit (ie the non-inv input of the op-amp) - so I'd try and make the wires on that pole as short as possible.
Which as short as possible?

**Reply by fhdm-dev** [reply]
Good question. I did the calculations a while ago so I can't remember exactly what method I used. I created a spreadsheet back then and so now I just plug in numbers until I get what I want.

Here is the spreadsheet: [https://docs.google.com/spreadsheets/d/1HvK-mElZuu1SMcotP0KdLk4wkiUVcHDVguAnWrdWbzg/edit?usp=sharing](https://docs.google.com/spreadsheets/d/1HvK-mElZuu1SMcotP0KdLk4wkiUVcHDVguAnWrdWbzg/edit?usp=sharing)

The idea is that you plug your resistor values into the white cells. The grey cells are the values calculated by the spreadsheet.
There's columns in the spreadsheet for 4 ranges and this assumes that R1 and R2 will be the same for each range. If that's not the case for your design just use the 'Range 0' column.

The one thing that makes the calculations relatively easy is that setting Rx to be equal to the parallel combination of R1 and R2 will give you a voltage range that is symmetrical around 0V.

The 0, 1, 2, 3 are just identifiers for each range, they don't represent voltage levels. In the app there is a screen where you can enter the min and max values for each range (tap the channel badge at the bottom of the screen and then tap Settings and Voltage Ranges). Your front-end "tells" the app which range is currently being used by applying appropriate (logic) high or low voltages to the voltage range GPIOs.

Regarding the position of the pins on the prototyping board, you can make them as close as you want. I don't think that will affect the noise.

**Comment by tyeth** [comment]
Super happy with this, thank you so much. Just built the first of three analog front ends...all went well apart from initial reversing the diodes 🤦 
![IMG_20230729_133629_925_15pc](https://github.com/fhdm-dev/scoppy/assets/6692083/7cf761f1-c5cc-42eb-a82d-9ddbbceaa3f8)
![IMG_20230729_183115_278_10pc](https://github.com/fhdm-dev/scoppy/assets/6692083/af97c09d-aa1f-4a7d-a445-5cc4d18c930c)
18v model for looking at my car, the backside of the board is embarrassing, but pleased I squeezed the caps as I did, and included the test point as double headers underneath (like CH1+2+GND)

**Reply by fhdm-dev** [reply]
Nice :smile:

**Comment by xDraconix** [comment]
Why dont use a smaler resistor value like 4.7k or 47k  instead off 470k?

Simulation of 50khz


![4 7k](https://github.com/fhdm-dev/scoppy/assets/28046753/b7b42395-6416-4e2f-b361-ccec3c79b588)
![470k](https://github.com/fhdm-dev/scoppy/assets/28046753/97e0d026-7126-442d-a231-9443e7c4ba86)

**Comment by fhdm-dev** [comment]
High resistor values are used to prevent the oscilloscope drawing too much current from the circuit under test.
I expect that adding small capacitors in parallel to R1, R2 and R4 would square up the waveform in your simulation.
eg. 22pf across R1 and R4 and 47pf across R2.

**Comment by Curt-is-Pi** [comment]
How important is it to use shielded test probes? I have planned to build a version of the design #5 using standard 4mm banana jack test leads. Probably wont use the setup for over 20khz.

**Comment by msalat** [comment]
The Pico samples with 500 kSa/s max → Nyquist → 250 kHz sin max. → 1.57 V/µs slew for ±1Vpp; for ±1.65 Vpp → 2.59 V/µs. The MCP6002 is only 0.6 V/µs. Perhaps OPA2340=6V/µs? But haven't checked input voltage range.
